### ISSUE TITLE
Audio readyState value "resetting" over time

### ISSUE THREAD

**Issue by obiot** [phoboslab/Ejecta#616]
Hi,

while playing with Howler, I noticed that when reusing a sound node, the corresponding `readyState` value of my audio assets are sometime changing (from 4 back to 2), which I did pinpoint after some debugging to the following line :
https://github.com/phoboslab/Ejecta/blob/master/Source/Ejecta/EJAudio/EJBindingAudio.m#L263

is there any reason here to set it (and also set it "back) to `kEJAudioHaveCurrentData` since Ejecta basically load all audio assets directly ? 

for now I fixed my issue by modifying the function as followed :
````objective-c
EJ_BIND_GET(readyState, ctx) {
	EJAudioReadyState state = kEJAudioHaveNothing;
	if( source ) {
        state = kEJAudioHaveEnoughData;
	}
	return JSValueMakeNumber(ctx, state);
}
````
but a better fix is certainly to keep the previous `readyState` value if already set to `4` (i.e. make it only incremental)

**Comment by phoboslab** [comment]
Help me interpret the spec here:
>> `HAVE_CURRENT_DATA` Data for the immediate current playback position is available, but either not enough data is available that the user agent could successfully advance the current playback position in the direction of playback at all without immediately reverting to the HAVE_METADATA state, or there is no more data to obtain in the direction of playback. For example, in video this corresponds to the user agent having data from the current frame, but not the next frame, when the current playback position is at the end of the current frame; and to **when playback has ended**.

~ https://dev.w3.org/html5/spec-preview/media-elements.html#dom-media-have_current_data

So this would mean the current behavior is actually correct. I'm not sure if this *"when playback has ended"* condition is overridden when the audio file is a) fully loaded and b) looped – which the current code also handles. Looped audio files should not reset to `HAVE_CURRENT_DATA`.

So, in short: I have no idea if, why and how the current code is misbehaving.

**Comment by obiot** [comment]
you are right, but basically the standard says that it should be changed back to `HAVE_CURRENT_DATA` if the current playback has ended and if the direction of playback is forward.

> A media element is said to have ended playback when:
>
>The element's readyState attribute is HAVE_METADATA or greater, and
>Either:
>
>The current playback position is the end of the media resource, and
>The direction of playback is forwards, and
>Either the media element does not have a loop attribute specified, or the media element has a current >media controller.
>Or:
>
>The current playback position is the earliest possible position, and
>The direction of playback is backwards.
>The ended attribute must return true if, the last time the event loop reached step 1, the media element >had ended playback and the direction of playback was forwards, and false otherwise.

which means that if `currentTime` is resetted to the beginning, it should then set back to `HAVE_ENOUGH_DATA`

**Comment by obiot** [comment]
to comply with the standard, what would make sense is to set back the `readyState` value here :
https://github.com/phoboslab/Ejecta/blob/master/Source/Ejecta/EJAudio/EJBindingAudio.m#L223
assuming that new value for `currentTime` is <= to the audio length (which would always be the case i'd say).

**Comment by phoboslab** [comment]
Good point! The above commits the `ended` flag when setting the `currentTime`. Does this fix your issue?

**Comment by obiot** [comment]
Hi, thanks !

I will try tomorrow, but will this prevent the `readyState` value to be set back to 2 ? 

unless i'm wrong this won't make the following (typical) use case to work, but maybe resetting the ended flag does the job :
```javascript
// play a sound for a while
sound.play()
...
// pause and restart from the beginning
sound.pause();
sound.currentTime = 0;
sound.play();
```

**Comment by obiot** [comment]
oh I see, the currentTime method is actually a getter, so yes it should prevent from changing back the value. I will test it tomorrow and let you know :)

thanks again !

**Comment by obiot** [comment]
sorry for the late feedback but I finally had the chance/time to test it against the latest version of Howler, and they now both work beautifully together :)

**Comment by phoboslab** [comment]
Cool, thanks for testing again!

**Comment by obiot** [comment]
Thank you as well for closing all the tickets I opened, it allowed me to remove all my temporary "hacks" in my development branch :):):)
